### ISSUE TITLE
package separator ' is not equivalent to :: when the package name starts with a NUL byte

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11246]
Migrated from [rt.perl.org#88138](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=88138) (status was 'resolved')

Searchable as RT88138$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



' and :&#8203;: should be interchangeable as package separators\.
They aren't for the corner case of a package name starting with a NUL byte&#8203;:

$ cat bless0\.pl 
\#\!/usr/bin/perl \-w
use strict;
use Devel&#8203;::Peek;

sub baz&#8203;::hello \{
&nbsp;    my $class = $\_\[0\];
&nbsp;    $class =~ s/\(\[^ \-~\]\)/sprintf "\\\\%03o"\, ord $1/ge;

&nbsp;    warn "Hello $class";
\}

foreach my $evil \(chr 1\, chr 0\) \{
&nbsp;    \{
&nbsp;        no strict 'refs';
&nbsp;        @&#8203;\{"foo&#8203;::$\{evil\}bar&#8203;::ISA"\} = 'baz';
&nbsp;    \}
&nbsp;    foreach my $sep \('&#8203;::'\, "'"\) \{
&nbsp;        my $o = bless \[\]\, 'foo' \. $sep \. $evil \. 'bar';
&nbsp;        $o\->hello;
&nbsp;    \}
\}

\_\_END\_\_
$ \./perl \-Ilib bless0\.pl 
Hello foo&#8203;::\\001bar=ARRAY\(0xa42a38\) at bless0\.pl line 9\.
Hello foo&#8203;::\\001bar=ARRAY\(0xa42a38\) at bless0\.pl line 9\.
Hello foo&#8203;::\\000bar=ARRAY\(0xa7ffd0\) at bless0\.pl line 9\.
Can't locate object method "hello" via package "foo'" at bless0\.pl line 19\.

\[It's the parsing loop at the start of Perl\_gv\_fetchpvn\_flags\(\)\]

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.14.0:

Configured by nick at Sat Apr  9 10:30:34 BST 2011.

Summary of my perl5 (revision 5 version 14 subversion 0) configuration:
  Derived from: d1ef1aa80747cb421af3e87bd1944a13a5fb2627
  Platform:
    osname=linux, osvers=2.6.35.4, archname=x86_64-linux
    uname='linux eris 2.6.35.4 #4 smp tue sep 21 09:54:22 bst 2010 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -DDEBUGGING -Doptimize=-Os -Uusethreads -Uuselongdouble -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.13.11-407-gd1ef1aa -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.13.11-407-gd1ef1aa -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -Umad -de'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    RC0


@INC for perl 5.14.0:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/site_perl/5.14.0/x86_64-linux
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/site_perl/5.14.0
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/5.14.0/x86_64-linux
    /home/nick/Sandpit/snap5.9.x-v5.13.11-379-g801eb08/lib/perl5/5.14.0
    .


Environment for perl 5.14.0:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Note to whoever fixes this \(maybe myself\)&#8203;:

This is probably just a matter of changing

&nbsp;        || \(\*name\_cursor == '\\'' && name\_cursor\[1\]\)\)
to

&nbsp;        || \(\*name\_cursor == '\\'' && name\_cursor \< name\_em1\)\)

or\, better yet\, changing

&nbsp;    if \(\(\*name\_cursor == '&#8203;:' && name\_cursor \< name\_em1
&nbsp;         && name\_cursor\[1\] == '&#8203;:'\)
&nbsp;        || \(\*name\_cursor == '\\'' && name\_cursor\[1\]\)\)

to

&nbsp;    if \(name\_cursor \< name\_em1 &&
&nbsp;        \(\(\*name\_cursor == '&#8203;:'
&nbsp;         && name\_cursor\[1\] == '&#8203;:'\)
&nbsp;        || \*name\_cursor == '\\''\)\)

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Now fixed with commit 46c0ec2\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
